internal/ethapi/api.go

```diff
@@ -778,27 +778,35 @@ func (s *PublicBlockChainAPI) EstimateGas(ctx context.Context, args CallArgs) (h
 	cap = hi
 
 	// Create a helper to check if a gas allowance results in an executable transaction
-	executable := func(gas uint64) bool {
+	executable := func(gas uint64) (bool, error) {
 		args.Gas = hexutil.Uint64(gas)
 
 		_, _, failed, err := s.doCall(ctx, args, rpc.PendingBlockNumber, 0)
 		if err != nil || failed {
-			return false
+			return false, err
 		}
-		return true
+		return true, nil
 	}
 	// Execute the binary search and hone in on an executable gas limit
 	for lo+1 < hi {
 		mid := (hi + lo) / 2
-		if !executable(mid) {
-			lo = mid
+		exec, execError := executable(mid)
+		if !exec {
+			if execError == vm.ErrOutOfGas { // Not enough gas
+				log.Debug("Ran out of gas, will increase lo to mid", "gas", mid)
+				lo = mid
+			} else { // Gas too high
+				log.Debug("Failed because gas too high, lowering hi to mid", "gas", mid)
+				hi = mid
+			}
 		} else {
 			hi = mid
 		}
 	}
 	// Reject the transaction as invalid if it still fails at the highest allowance
 	if hi == cap {
-		if !executable(hi) {
+		exec, _ := executable(hi)
+		if !exec {
 			return 0, fmt.Errorf("gas required exceeds allowance or always failing transaction")
```

Comment:
We can change this comment to "gas required for transaction exceeds allowance" because we will now only reach this point on consistent "out of gas" errors.

Reply:
Good catch, updated comment